File: C++/1040-longest_symmetric_string.cc
#include <iostream>
#include <vector>
#include <queue>


struct SubStr {
    std::size_t start, end;
};


SubStr longest_symmetric_string(std::string &s) {
    std::queue<SubStr> Queue;
    // odd substr
    for (std::size_t i=0; i<s.size(); ++i) {
        Queue.push({i, i});
    }
    // even substr
    for (std::size_t i=0; i<s.size()-1; ++i) {
        if (s[i] == s[i+1]) {
            Queue.push({i, i+1});
        }
    }
    SubStr substr;
    while (!Queue.empty()) {
        substr = Queue.front();
        Queue.pop();
        if (substr.start > 0 && substr.end < s.size() - 1 &&
            s[substr.start-1] == s[substr.end+1])
        {
            Queue.push({substr.start-1, substr.end+1});
        }
    }
    return substr;
}


int main() {

    std::string s;
    std::getline(std::cin, s);

    SubStr res = longest_symmetric_string(s);
    std::cout << res.end - res.start + 1 << '\n';

    return 0;
}

File: Python/1040-longest_symmetric_string.py
from collections import namedtuple, deque


Substr = namedtuple('Substr', ['start', 'end'])


def longest_symmetric_string(s):
    queue = deque()
    # odd substr
    for i in range(len(s)):
        queue.append(Substr(i, i))
    # even substr
    for i in range(len(s) - 1):
        if s[i] == s[i+1]:
            queue.append(Substr(i, i+1))
    while queue:
        substr = queue.popleft()
        if (substr.start > 0 and substr.end < len(s) - 1 and
                s[substr.start-1] == s[substr.end+1]):
            queue.append(Substr(substr.start-1, substr.end+1))
    return substr


if __name__ == '__main__':
    s = input()
    start, end = longest_symmetric_string(s)
    print(end - start + 1)
